src/main/scala/mesosphere/marathon/metrics/SlidingAverageSnapshot.scala

```diff
@@ -0,0 +1,163 @@
+    }
+
+    // Create an array with that many free slots, as the full ticks that fit within the duration requested
+    logger.info(s"Initialized sliding average reporter with ${fullFrames} frames")
+    Array.fill(fullFrames){ null }
+  }
+
+  /**
+    * Combine all the metrics in the ring buffer and return an
+    * @return
+    */
+  private def combineRingMetrics: TickMetricSnapshot = {
+    var rangeFrom: MilliTimestamp = MilliTimestamp.now
+    var rangeTo: MilliTimestamp = MilliTimestamp.now
+    var combined: Option[Map[Entity, EntitySnapshot]] = None
+
+    // To combine the metrics we need to iterate over the ring buffer, performing
+    // the following operations:
+    //
+    // - Find the oldest and the newest timestamp, that will be used as the snapshot
+    //   range for our combined metrics.
+    // - For every snapshot, merge every metric with the current
+    //
+    // Since we are putting items in the ring buffer in order, we can exploit the indices
+    // to quickly identify the oldest and the newest entry:
+    //
+    // - Since we post-increment when adding items on the ring buffer, the item with
+    //   index `averageIndex - 1` is the newest one.
+    // - Consequently, the item with index `averageIndex` is the oldest one.
+    //
+    // In the following loop we are starting from the oldest item and walk our way to the
+    // newest item. That's why we start with `idx = averageIndex` and we advance it
+    // as many times as the items in the ring buffer, effectively reaching to the
+    // newest item on `idx = averageIndex - 1`
+    //
+
+    val maxItems: Int = averageRing.length
+    var i: Int = 0
+    while (i < maxItems) {
+      val idx = (averageIndex + i) % maxItems
+      val inst = averageRing(idx)
+
+      //
+      // The `null` items are items not yet initialized and we
+      // therefore consider them transparent.
+      //
+      combined = inst match {
+        case null => combined
+        case _ => {
+          combined match {
+
+            //
+            // The first record we find is the oldest one. So we are using it's
+            // timestamp as the `rangeFrom` and we initialize the `combined`
+            // snapshot with it's metrics.
+            //
+            // Just in case this is the last item we ever find, we are also
+            // populating `rangeTo` with the correct time bounds.
+            //
+            case None => {
+              rangeFrom = inst.from
+              rangeTo = inst.to
+              Some(inst.metrics)
+            }
+
+            //
+            // Otherwise combine the current metrics and advance `rangeTo`
+            //
+            case Some(current) => {
+              rangeTo = inst.to
+              Some(MapMerge.Syntax(current).merge(inst.metrics, (l, r) => l.merge(r, collectionContext)))
+            }
+          }
+        }
+      }
+
+      // Advance index of while loop
+      i += 1
+    }
```

Comment:
Thanks for the docs. I get it now 🙂. How about encapsulating the iteration in an iterator?

```scala

/**
 * Iterates from oldest to news item in a ring buffer.
 */
class RingBufferIterator(buffer: Array[TickMetricSnapshot]) extends Iterator[TickMetricSnapshot] {
  var i: Int = 0
  val maxItems: Int = buffer.length
   
  def getNextIndext() = (averageIndex + i) % maxItems

  override def hasNext(): Boolean = {
    val idx = getNextIndext()
    val inst = averageRing(idx)
    if (inst == null) return false
    if (idx == averageIndex) return false // we made a full round
    return true
  }

  override def next(): TickMetricSnapshot = {
    val idx = getNextIndext()
    i += 1
    averageRing(idx)
  }
}
```

The averaging would then be

```scala
val iter = RingBufferIterator(averageRing)
val oldest = iter.next()
val rangeFrom = oldest.from

iter.foldLeft( (oldest.metrics, oldest.to)) { case ((combined, _), next) =>
  val updated = MapMerge.Syntax(next).merge(combined.metrics, (l, r) => l.merge(r, collectionContext))
  (updated, next.to)
}
```